### DISCUSSION TITLE
How does Next SEO with Sanity or any headless CMS work?

### DISCUSSION THREAD

**Discussion by mikhael28** [vercel/next.js#18512] [Help] [answered]
Hi all, I am a little confused about how a headless CMS like Sanity.io, combined with Next.js, when deployed would have good SEO for blog posts. For example, let's say I created a blog post on Sanity and run through the build process with Next, and deploy. How does Google know that I have a blog post called 'How to write great GitHub tickets' on my Next site? 

Does it load the JavaScript in an environment of some sort, actually execute the fetch requests, then crawls through the now populated blog posts? And then indexes the resulting route tree, even though there aren't multiple routes - it's really a re-usable component that tricks google by putting in unique metadata from the blog post?

That makes sense... but it almost sounds too good to be true. Maybe I'm under-estimating how good a job Google has done in improving how they index websites, and now does not require a technical 'file' - they can auto-create indexes through dynamic routes?

I guess my question is more about how SEO works with dynamic React routing - if anyone can confirm my suspicion here, that would be awesome. I'm sorry it's not worded exceedingly well, I hope people can follow my train of thought.

**Comment by lfades** [comment, marked as answer]
> How does Google know that I have a blog post called 'How to write great GitHub tickets' on my Next site?

Google crawls public websites, and also follows links to discover more pages.

> Does it load the JavaScript in an environment of some sort, actually execute the fetch requests, then crawls through the now populated blog posts?

Google does that, yes. Not all crawlers do it, but in any case it shouldn't be an issue if the content is static HTML (using `getStaticProps`). JS won't need to run because the data is already populated by the initial HTML.

> it's really a re-usable component that tricks google by putting in unique metadata from the blog post?

I'm not sure I understand but here are my 2 cents: Crawlers find your blog posts because they find links to them, usually in your blog, which is usually linked your home (`/`), they also follow links in social media, so if you share the URL of your blog they'll know about it.

> they can auto-create indexes through dynamic routes?

Same answer as above, those crawlers are smart.

**Comment by sakshiiiqw** [comment]
Next SEO with Sanity or any headless CMS works by combining a fast frontend framework like Next.js with centralized content management and optimized metadata handling for search engines. In this setup, content is managed inside the CMS while the frontend fetches and displays it dynamically. Tools like Next SEO help developers manage meta titles, descriptions, Open Graph tags, structured data, and canonical URLs more efficiently, which improves visibility on search engines. This approach is popular because it offers better performance, scalability, and flexibility compared to traditional CMS platforms. Many digital and mobile growth teams, including Microbit Media, often explore headless CMS architectures to create faster user experiences that help businesses connect and organize people more effectively across web and mobile platforms.
